src/shims/ffi_support.rs

```diff
@@ -0,0 +1,283 @@
+    /// and convert it to a `CArg`.
+    fn scalar_to_carg(
+        k: ScalarMaybeUninit<Tag>,
+        arg_type: &Ty<'tcx>,
+        cx: &impl HasDataLayout,
+    ) -> InterpResult<'tcx, CArg> {
+        match arg_type.kind() {
+            // If the primitive provided can be converted to a type matching the hir type pattern
+            // then create a `CArg` of this primitive value with the corresponding `CArg` constructor.
+            // the ints
+            TyKind::Int(IntTy::I8) => {
+                return Ok(CArg::Int8(k.to_i8()?));
+            }
+            TyKind::Int(IntTy::I16) => {
+                return Ok(CArg::Int16(k.to_i16()?));
+            }
+            TyKind::Int(IntTy::I32) => {
+                return Ok(CArg::Int32(k.to_i32()?));
+            }
+            TyKind::Int(IntTy::I64) => {
+                return Ok(CArg::Int64(k.to_i64()?));
+            }
+            TyKind::Int(IntTy::Isize) => {
+                return Ok(CArg::ISize(k.to_machine_isize(cx)?.try_into().unwrap()));
+            }
+            // the uints
+            TyKind::Uint(UintTy::U8) => {
+                return Ok(CArg::UInt8(k.to_u8()?));
+            }
+            TyKind::Uint(UintTy::U16) => {
+                return Ok(CArg::UInt16(k.to_u16()?));
+            }
+            TyKind::Uint(UintTy::U32) => {
+                return Ok(CArg::UInt32(k.to_u32()?));
+            }
+            TyKind::Uint(UintTy::U64) => {
+                return Ok(CArg::UInt64(k.to_u64()?));
+            }
+            TyKind::Uint(UintTy::Usize) => {
+                return Ok(CArg::USize(k.to_machine_usize(cx)?.try_into().unwrap()));
+            }
+            _ => {}
+        }
+        // If no primitives were returned then we have an unsupported type.
+        throw_unsup_format!(
+            "unsupported scalar argument type to external C function: {:?}",
+            arg_type
+        );
+    }
+
+    /// Call external C function and
+    /// store output, depending on return type in the function signature.
+    fn call_external_c_and_store_return<'a>(
+        &mut self,
+        external_fct_defn: ExternalCFuncDeclRep<'tcx>,
+        dest: &PlaceTy<'tcx, Tag>,
+        ptr: CodePtr,
+        libffi_args: Vec<libffi::high::Arg<'a>>,
+    ) -> InterpResult<'tcx, ()> {
+        let this = self.eval_context_mut();
+
+        // Unsafe because of the call to external C code.
+        // Because this is calling a C function it is not necessarily sound,
+        // but there is no way around this and we've checked as much as we can.
+        unsafe {
+            // If the return type of a function is a primitive integer type,
+            // then call the function (`ptr`) with arguments `libffi_args`, store the return value as the specified
+            // primitive integer type, and then write this value out to the miri memory as an integer.
+            match external_fct_defn.output_type.kind() {
```

Comment:
Hmm... can we avoid the entire `fn_sig` business and just use `dest.layout.ty` here?